### PR TITLE
Improvements to pages plugin

### PR DESCRIPTION
**Pull request by hypeJunction**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -36,7 +36,8 @@
         "misd/linkify": "~1.1.2",
         "league/flysystem-memory": "^1.0",
         "htmlawed/htmlawed": "^1.1.22",
-        "imagine/imagine": "^0.6.3"
+        "imagine/imagine": "^0.6.3",
+        "bower-asset/jquery-treeview": "^1.4"
```
Comment (jdalsem) [inline composer.json]: no like... i know why you did this, but still i do not like it... can accept for now, but in an ideal world this should stay with the plugin composer..

**Review by jdalsem — APPROVED** [review]

File: composer.json (hunk 1 above)
Comment (hypeJunction) [inline composer.json]: Yes, ideally, pages plugin would live in its own repo

File: mod/pages/start.php [hunk 2]
```diff
@@ -56,7 +48,8 @@ function pages_init() {
 		'title' => 'text',
 		'description' => 'longtext',
 		'tags' => 'tags',
-		'parent_guid' => 'parent',
+		'categries' => 'categories',
```
Comment (hypeJunction) [inline mod/pages/start.php]: typo

File: mod/pages/views/default/resources/pages/view.php [hunk 3]
```diff
@@ -14,46 +14,47 @@
...
+
+$title = $page->getDisplayName();
 elgg_push_breadcrumb($title);
 
-$content = elgg_view_entity($page, [
-	'full_view' => true,
-	'show_responses' => true,
-]);
+$content = elgg_view_entity($page, array('full_view' => true));
+$content .= elgg_view_comments($page);
```
Comment (hypeJunction) [inline mod/pages/views/default/resources/pages/view.php]: Should go into `responses`

File: mod/pages/lib/pages.php [hunk 4]
```diff
@@ -108,44 +111,51 @@ function pages_get_navigation_tree($container) {
...
  * @param ElggEntity $container Container entity for the pages
+ * @param ElggEntity $selected  Selected page
+ *
+ * @return void
  */
-function pages_register_navigation_tree($container) {
+function pages_register_navigation_tree($container, $selected = null) {
 	$pages = pages_get_navigation_tree($container);
+
 	if ($pages) {
```
Comment (jeabakker) [inline mod/pages/lib/pages.php]: is a quick return possible here?

File: mod/pages/views/default/pages/input/parent.php [hunk 5]
```diff
@@ -1,37 +1,6 @@
 <?php
 /**
- * Parent picker
- *
- * @uses $vars['value']          The current value, if any
- * @uses $vars['options_values']
- * @uses $vars['name']           The name of the input field
- * @uses $vars['entity']         Optional. The child entity (uses container_guid)
+ * @depreated 3.0 Use input/pages/parent
  */
```
Comment (jeabakker) [inline mod/pages/views/default/pages/input/parent.php]: maybe also add a deprecation notice so devs actually change the use of this view

File: mod/pages/views/default/resources/pages/edit.php [hunk 6]
```diff
@@ -1,44 +1,53 @@
...
-	
-	$content = elgg_view_form('pages/edit', [], $vars);
-} else {
-	$content = elgg_echo("pages:noaccess");
-}
+$vars = pages_prepare_form_vars($page, $page->parent_guid);
+$content = elgg_view_form('pages/edit', [], $vars);
 
-$body = elgg_view_layout('content', [
+$body = elgg_view_layout('content', array(
```
Comment (jeabakker) [inline mod/pages/views/default/resources/pages/edit.php]: shouldn't this be a short array eg `[]`

File: mod/pages/views/default/resources/pages/view.php [hunk 7]
```diff
@@ -14,46 +14,47 @@
...
 pages_prepare_parent_breadcrumbs($page);
+
+$title = $page->getDisplayName();
 elgg_push_breadcrumb($title);
 
-$content = elgg_view_entity($page, [
-	'full_view' => true,
-	'show_responses' => true,
-]);
+$content = elgg_view_entity($page, array('full_view' => true));
```
Comment (jeabakker) [inline mod/pages/views/default/resources/pages/view.php]: short array notation

File: mod/pages/views/default/resources/pages/view.php [hunk 8]
```diff
@@ -14,46 +14,47 @@
...
+$content .= elgg_view_comments($page);
 
 // can add subpage if can edit this page and write to container (such as a group)
 if ($page->canEdit() && $container->canWriteToContainer(0, 'object', 'page')) {
-	$url = "pages/add/$page->guid";
-	elgg_register_menu_item('title', [
+	elgg_register_menu_item('title', array(
```
Comment (jeabakker) [inline mod/pages/views/default/resources/pages/view.php]: short array notation

File: mod/pages/views/default/resources/pages/view.php [hunk 9]
```diff
@@ -14,46 +14,47 @@
...
-			'href' => $url,
+			'href' => "pages/add/$page->guid",
 			'text' => elgg_echo('pages:newchild'),
 			'link_class' => 'elgg-button elgg-button-action',
-	]);
+	));
 }
 
-$body = elgg_view_layout('content', [
+$body = elgg_view_layout('content', array(
```
Comment (jeabakker) [inline mod/pages/views/default/resources/pages/view.php]: short array notation

File: mod/pages/views/default/input/pages/parent.php [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+foreach ($pages as $page) {
+	$spacing = "";
+	for ($i = 0; $i < $page['depth']; $i++) {
+		$spacing .= "--";
+	}
+	$options[$page['guid']] = "$spacing " . $page['title'];
+}
+
+$defaults = array(
```
Comment (jeabakker) [inline mod/pages/views/default/input/pages/parent.php]: short array notation

**Review by jeabakker — CHANGES_REQUESTED** [review]

File: composer.json (hunk 1 above)
Comment (mrclay) [inline composer.json]: Can this be brought in via package.json? (and ideally all the others)

Comment (mrclay) [inline composer.json]: (...but I'm OK with this here)

**Review by jeabakker — APPROVED** [review]

File: mod/pages/views/default/pages/input/parent.php (hunk 5 above)
Comment (jdalsem) [inline mod/pages/views/default/pages/input/parent.php]: also deprecated is typed incorrectly

**Comment by hypeJunction** [comment]
Thinking out loud: 
I think we should just loose the treeview and use our menu API. However, looking at the menu API, it feels that the whole `child_menu` approach isn't all that great. I propose we loose it, and instead add menu templates: `tree`, `navbar`

File: mod/pages/views/default/pages/input/parent.php (hunk 5 above)
Comment (jdalsem) [inline mod/pages/views/default/pages/input/parent.php]: by the way... why not just remove the view? document in upgrade and it is gone

File: mod/pages/views/default/resources/pages/edit.php [hunk 11]
```diff
@@ -1,44 +1,53 @@
...
 }
 
 $container = $page->getContainerEntity();
 if (!$container) {
 	register_error(elgg_echo('noaccess'));
-	forward('');
+	forward('', '403');
+}
+
+elgg_set_page_owner_guid($container->guid);
```
Comment (jdalsem) [inline mod/pages/views/default/resources/pages/edit.php]: doesn't this need to happen in the page handler?

File: mod/pages/views/default/resources/pages/view.php [hunk 12]
```diff
@@ -14,23 +14,24 @@
...
-
-$container = elgg_get_page_owner_entity();
+$container = $page->getContainerEntity();
 if (!$container) {
-	forward(REFERER);
+	forward('', '404');
 }
 
-$title = $page->title;
+elgg_set_page_owner_guid($container->guid);
```
Comment (jdalsem) [inline mod/pages/views/default/resources/pages/view.php]: doesn't this need to happen in the page handler?

**Review by jdalsem — APPROVED** [review]

**Comment by jdalsem** [comment]
> I think we should just loose the treeview and use our menu API.

also was thinking something like that... but let us finish this PR first... can remove more easier after that.... there is also a benefit to the jstree view functionality like drag/drop reordering of pages and all kinds of events

File: mod/pages/views/default/resources/pages/edit.php (hunk 11 above)
Comment (hypeJunction) [inline mod/pages/views/default/resources/pages/edit.php]: Why? As far as I am concerned nothing should be happening in the page handler - they should be gone altogether.

Comment (jdalsem) [inline mod/pages/views/default/resources/pages/edit.php]: can discuss this somewhere else

**Comment by jdalsem** [comment]
if you just remove the deprecated view this can be merged

**Comment by hypeJunction** [comment]
It's not bothering me, and I have spent way too much time on this as it is

**Comment by jdalsem** [comment]
will remove it in a following PR

**Pull request merged**